editoast/src/views/timetable/stdcm.rs

```diff
@@ -561,37 +685,56 @@ fn build_single_margin(margin: Option<MarginValue>) -> Margins {
     }
 }
 
-/// Build the list of work schedules for the given time range
-async fn build_work_schedules(
-    conn: &mut DbConnection,
-    time: DateTime<Utc>,
+fn filter_core_work_schedule(
+    ws: &WorkSchedule,
+    start_time: DateTime<Utc>,
+) -> crate::core::stdcm::WorkSchedule {
+    crate::core::stdcm::WorkSchedule {
+        start_time: elapsed_since_time_ms(&ws.start_date_time, &start_time),
+        end_time: elapsed_since_time_ms(&ws.end_date_time, &start_time),
+        track_ranges: ws
+            .track_ranges
+            .iter()
+            .map(|track| UndirectedTrackRange {
+                track_section: track.track.to_string(),
+                begin: (track.begin * 1000.0) as u64,
+                end: (track.end * 1000.0) as u64,
+            })
+            .collect(),
+    }
+}
+
+fn filter_stdcm_work_schedules(
+    work_schedules: &[WorkSchedule],
+    start_time: DateTime<Utc>,
     maximum_run_time: u64,
-    work_schedule_group_id: i64,
-) -> Result<Vec<CoreWorkSchedule>> {
-    let selection_setting: SelectionSettings<WorkSchedule> = SelectionSettings::new()
-        .filter(move || WorkSchedule::WORK_SCHEDULE_GROUP_ID.eq(work_schedule_group_id));
-    let res = Ok(WorkSchedule::list(conn, selection_setting)
-        .await?
+) -> Vec<crate::core::stdcm::WorkSchedule> {
+    work_schedules
         .iter()
-        .map(|ws| {
-            let schedule = CoreWorkSchedule {
-                start_time: elapsed_since_time_ms(&ws.start_date_time, &time),
-                end_time: elapsed_since_time_ms(&ws.end_date_time, &time),
-                track_ranges: ws
-                    .track_ranges
-                    .iter()
-                    .map(|track| UndirectedTrackRange {
-                        track_section: track.track.to_string(),
-                        begin: (track.begin * 1000.0) as u64,
-                        end: (track.end * 1000.0) as u64,
-                    })
-                    .collect(),
-            };
-            schedule
-        })
+        .map(|ws| filter_core_work_schedule(ws, start_time))
         .filter(|ws| ws.end_time > 0 && ws.start_time < maximum_run_time)
-        .collect());
-    res
+        .collect()
+}
+
+fn filter_conflict_work_schedules(
+    work_schedules: &[WorkSchedule],
+    start_time: DateTime<Utc>,
+    maximum_run_time: u64,
+) -> Option<WorkSchedulesRequest> {
+    if work_schedules.is_empty() {
+        return None;
+    }
+
+    let work_schedule_requirements = work_schedules
+        .iter()
+        .map(|ws| (ws.id, filter_core_work_schedule(ws, start_time)))
```

Comment:
Since it's actually 2 steps, I find it better to write it as 2 steps, just fewer parents nesting too.

```suggestion
        .map(|ws| filter_core_work_schedule(ws, start_time))
        .map(|filtered| (ws.id, filtered))
```

Reply:
I can't do the two steps because `filter_core_work_schedule` returns a struct without `id`.

Reply:
🤦